File: src/g2t.cpp

#include <numeric>
#include <algorithm>
#include <iostream>
#include <memory>
#include <vector>

#include <IITree.h>
#include "types.h"
#include "bramble.h"
#include "evaluate.h"
#include "g2t.h"

extern bool SOFT_CLIPS;
extern bool STRICT;
extern bool USE_FASTA;
extern bool LONG_READS;

extern GFastaDb *gfasta;  // FASTA file, -S

namespace bramble {

  IITData::IITData() {}

  IntervalTree::IntervalTree() {
    iit = new IITree<int, IITData>();
  }

  IntervalTree::~IntervalTree() { 
    delete(iit); 
  }

  // Add guide exon with TID and transcript start
  void IntervalTree::addInterval(const tid_t &tid, IntervalData interval, 
                                const char* ref_name) {
    IITData node = IITData();
    node.tid = tid;
    node.exon_id = interval.idx;
    node.pos_start = interval.pos_start;

    node.has_prev = interval.has_prev;
    node.has_next = interval.has_next;
    node.prev_start = interval.prev_start;
    node.prev_end = interval.prev_end;
    node.next_start = interval.next_start;
    node.next_end = interval.next_end;
    node.transcript_len = interval.transcript_len;

    // grab here to avoid multithreading issues
    if (USE_FASTA) {
      GFaSeqGet* faseq = gfasta->fetch(ref_name);
      const char *seq = faseq->copyRange(interval.start, 
        interval.end - 1, false, true);
      node.seq = seq;
    }

    iit->add(interval.start, interval.end, node);
  }

  void IntervalTree::indexTree() {
    iit->index();
  }

  struct iitData {
    uint32_t start;
    uint32_t end;
    IITData data;
  };

  bool
  IntervalTree::findOverlappingForTid(uint32_t qstart, uint32_t qend, tid_t tid,
                                      GuideExon &gexon) {
    
    if (qstart == 0 && qend == 0) return false;
    std::vector<size_t> hits;
    iit->overlap(qstart, qend, hits);

    for (size_t idx : hits) {
      uint32_t s = iit->start(idx);
      uint32_t e = iit->end(idx);

      const IITData& data = iit->data(idx);
      if (data.tid == tid) {
        gexon.start = s;
        gexon.end = e;
        gexon.transcript_len = data.transcript_len;
        gexon.pos_start = data.pos_start;
        gexon.exon_id = data.exon_id;
        gexon.has_prev = data.has_prev;
        gexon.has_next = data.has_next;
        gexon.prev_start = data.prev_start;
        gexon.prev_end = data.prev_end;
        gexon.next_start = data.next_start;
        gexon.next_end = data.next_end;
        gexon.seq = data.seq; // may have to be cleared
        return true;
      }
    }

    return false;
  }

  bool
  IntervalTree::findOverlapping(uint32_t qstart, uint32_t qend, 
                                char strand, ReadEvaluationConfig config, 
                                ExonStatus status, 
                                std::vector<GuideExon> &gexons) {
    std::vector<size_t> hits;
    
    iit->overlap(qstart, qend, hits);
    if (hits.empty()) return false;

    uint32_t pos;
    uint32_t left_gap;
    uint32_t left_ins;
    uint32_t right_gap;
    uint32_t right_ins;

    // remove tids that appear >1
    // means that junctions have been ignored

    // update: should be handled in create_match loop
    // if (LONG_READS) { // won't happen with strict short read parameters
    //   thread_local unordered_set<tid_t> seen;
    //   seen.clear();
    //   thread_local unordered_set<tid_t> duplicates;
    //   duplicates.clear();

    //   for (size_t idx : hits) {
    //     tid_t tid = iit->data(idx).tid;
    //     if (!seen.emplace(tid).second) {
    //       duplicates.emplace(tid);
    //     }
    //   }

    //   hits.erase(
    //     std::remove_if(hits.begin(), hits.end(), [&](size_t idx) {
    //       return duplicates.count(iit->data(idx).tid);
    //     }),
    //     hits.end()
    //   );
    // }

    for (size_t idx : hits) {
      uint32_t s = iit->start(idx);
      uint32_t e = iit->end(idx);

      pos = 0;
      left_gap = 0;
      left_ins = 0;
      right_gap = 0;
      right_ins = 0;

      const IITData& data = iit->data(idx);

      if (strand == '+') {
        if (s <= qstart) {
          // Left junction gap: query starts after interval starts
          pos = (qstart - s) + data.pos_start;
          left_gap = qstart - s;
          if (status == MIDDLE_EXON || status == LAST_EXON) {
            if (left_gap > config.max_junc_gap) continue;
          }
        } else {
          pos = data.pos_start;
          // Left clip: query extends before interval
          left_ins = s - qstart;
          if (status == MIDDLE_EXON || status == LAST_EXON) {
            if (left_ins > config.max_ins) continue;
          } else {
            if (left_ins > config.max_clip) continue;
          }
        }

        // Check right side
        if (e < qend) {
          // Right clip: query extends past interval
          right_ins = qend - e;
          if (status == FIRST_EXON || status == MIDDLE_EXON) {
            if (right_ins > config.max_ins) continue;
          } else {
            if (right_ins > config.max_clip) continue;
          }
        } else if (qend < e) {
          // Right junction gap: query ends before interval ends
          right_gap = e - qend;
          if (status == FIRST_EXON || status == MIDDLE_EXON) {
            if (right_gap > config.max_junc_gap) continue;
          }
        }

      } else {  // strand == '-'
        if (qend <= e) {
          pos = (e - qend) + data.pos_start;
          right_gap = e - qend;
          // Right junction gap: query ends before interval ends
          if (status == FIRST_EXON || status == MIDDLE_EXON) {
            if (right_gap > config.max_junc_gap) continue;
          }
        } else {
          pos = data.pos_start;
          // Right clip: query extends past interval
          right_ins = qend - e;
          if (status == FIRST_EXON || MIDDLE_EXON) {
            if (right_ins > config.max_ins) continue;
          } else {
            if (right_ins > config.max_clip) continue;
          }
        }

        // Check left side
        if (qstart < s) {
          // Left clip: query extends before interval
          left_ins = s - qstart;
          if (status == MIDDLE_EXON || status == LAST_EXON) {
            if (left_ins > config.max_ins) continue;
          } else {
            if (left_ins > config.max_clip) continue;
          }
        } else if (s < qstart) {
          left_gap = qstart - s;
          // Left junction gap: query starts after interval starts
          if (status == MIDDLE_EXON || status == LAST_EXON) {
            if (left_gap > config.max_junc_gap) continue;
          }
        }
      }

      auto exon = GuideExon();
      exon.start = s;
      exon.end = e;
      exon.tid = data.tid;

      exon.pos = pos;
      exon.left_gap = left_gap;
      exon.left_ins = left_ins;
      exon.right_gap = right_gap;
      exon.right_ins = right_ins;

      exon.pos_start = data.pos_start;
      exon.exon_id = data.exon_id;

      exon.has_prev = data.has_prev;
      exon.has_next = data.has_next;
      exon.prev_start = data.prev_start;
      exon.prev_end = data.prev_end;
      exon.next_start = data.next_start;
      exon.next_end = data.next_end;

      exon.transcript_len = data.transcript_len;
      exon.seq = data.seq; // may have to be cleared

      gexons.emplace_back(exon);
    }

    return (!gexons.empty());
  }

  g2tTree::g2tTree() = default;
    
  g2tTree::~g2tTree() {
    // Clean up all allocated trees
    for (auto& pair : trees) {
      delete pair.first;   // forward tree
      delete pair.second;  // reverse tree
    }
    trees.clear();
  }

  // Get or create tree pair for a given refid
  void g2tTree::createTree(int refid) {
    // Create new tree pair for this refid
    IntervalTree* fw_tree = new IntervalTree();
    IntervalTree* rc_tree = new IntervalTree();
    trees.emplace_back(std::make_pair(fw_tree, rc_tree));
  }

  IntervalTree *g2tTree::getTreeForStrand(int refid, char strand) {
    if (refid < 0 || (size_t)refid >= trees.size()) return nullptr;
    auto pair = trees[refid];
    if (strand == '+' || strand == 1)
      return pair.first;
    else if (strand == '-' || strand == -1)
      return pair.second;
    else
      return nullptr;
  }

  // If tid_name is already part of the tree, then return the existing
  // tid for it, otherwise give it the next available tid and return that.
  tid_t g2tTree::insertTidString(const char *&tid_name, BamIO* io) {
    tid_t tid = io->get_tid(tid_name);

    auto lookup = name_id_map.find(tid);
    if (lookup == name_id_map.end()) {
      name_id_map.insert({tid, tid_name});
      tid_names.push_back(tid_name);
    }

    return tid;
  }

  // get the string name given an id
  const std::string &g2tTree::getTidName(tid_t id) {
    return (name_id_map.find(id) != name_id_map.end()) ? 
      name_id_map[id] : invalid_name;
  }

  // Add guide exon with TID and transcript start
  void g2tTree::addInterval(refid_t refid, const tid_t &tid, 
                            IntervalData interval, char strand,
                            const char* ref_name) {
    IntervalTree *tree = getTreeForStrand(refid, strand);
    tree->addInterval(tid, interval, ref_name);
  }

  // Index tree after guides have been added
  void g2tTree::indexTrees(int refid) {
    IntervalTree *fw_tree = getTreeForStrand(refid, '+');
    IntervalTree *rc_tree = getTreeForStrand(refid, '-');
    fw_tree->indexTree();
    rc_tree->indexTree();
  }

  bool
  g2tTree::getGuideExonForTid(int refid, char strand, tid_t tid,
                              uint32_t start, uint32_t end, GuideExon &gexon) {
    IntervalTree *tree = getTreeForStrand(refid, strand);
    if (!tree) return false;
    return tree->findOverlappingForTid(start, end, tid, gexon);
  }

  // Find all guide TIDs that overlap with a read exon
  bool
  g2tTree::getGuideExons(int refid, char strand, GSeg exon,
                        ReadEvaluationConfig config, ExonStatus status,
                        std::vector<GuideExon> &gexons) {
    IntervalTree *tree = getTreeForStrand(refid, strand);
    if (!tree) {
      return false;
    }
    return tree->findOverlapping(exon.start, exon.end, strand, 
      config, status, gexons);
  }
  
};

File: bramble-rs/src/g2t.rs
use crate::annotation::{Exon, Transcript};
use crate::evaluate::{ExonStatus, ReadEvaluationConfig};
use crate::fasta::FastaDb;
use crate::types::{HashMap, HashMapExt, HashSet, HashSetExt, RefId, Tid};
use anyhow::{anyhow, Result};
use coitrees::{BasicCOITree, Interval, IntervalTree as CoitreeIntervalTree};
use std::sync::Arc;

#[derive(Debug, Clone)]
pub struct IntervalData {
    pub start: u32,
    pub end: u32,
    pub idx: u8,
    pub pos_start: u32,
    pub seq: Arc<[u8]>,

    pub has_prev: bool,
    pub has_next: bool,
    pub prev_start: u32,
    pub prev_end: u32,
    pub next_start: u32,
    pub next_end: u32,

    pub transcript_len: u32,
}

#[derive(Debug, Clone)]
pub struct GuideExon {
    pub start: u32,
    pub end: u32,
    pub pos: u32,
    pub pos_start: u32,
    pub exon_id: u8,
    pub seq: Arc<[u8]>,
    pub left_ins: i32,
    pub right_ins: i32,
    pub left_gap: i32,
    pub right_gap: i32,

    pub has_prev: bool,
    pub has_next: bool,
    pub prev_start: u32,
    pub prev_end: u32,
    pub next_start: u32,
    pub next_end: u32,

    pub transcript_len: u32,
}

#[derive(Debug, Clone, Default)]
pub struct IITData {
    pub tid: Tid,
    pub exon_id: u8,
    pub pos_start: u32,
    pub has_prev: bool,
    pub has_next: bool,
    pub prev_start: u32,
    pub prev_end: u32,
    pub next_start: u32,
    pub next_end: u32,
    pub transcript_len: u32,
}

pub struct IntervalTree {
    intervals: Vec<Interval<IITData>>,
    tree: Option<BasicCOITree<IITData, u32>>,
    seqs: HashMap<(Tid, u8), Arc<[u8]>>,
}

impl IntervalTree {
    pub fn new() -> Self {
        Self {
            intervals: Vec::new(),
            tree: None,
            seqs: HashMap::new(),
        }
    }

    pub fn add_interval(&mut self, tid: Tid, interval: &IntervalData) {
        let data = IITData {
            tid,
            exon_id: interval.idx,
            pos_start: interval.pos_start,
            has_prev: interval.has_prev,
            has_next: interval.has_next,
            prev_start: interval.prev_start,
            prev_end: interval.prev_end,
            next_start: interval.next_start,
            next_end: interval.next_end,
            transcript_len: interval.transcript_len,
        };
        if !interval.seq.is_empty() {
            self.seqs.insert((tid, interval.idx), Arc::clone(&interval.seq));
        }
        // COITree intervals are end-inclusive; convert [start, end) -> [start, end-1].
        let first = interval.start as i32;
        let last = interval.end.saturating_sub(1) as i32;
        if last >= first {
            self.intervals.push(Interval::new(first, last, data));
        }
    }

    pub fn index(&mut self) {
        self.tree = Some(BasicCOITree::new(&self.intervals));
    }

    pub fn find_overlapping_for_tid(
        &self,
        qstart: u32,
        qend: u32,
        tid: Tid,
    ) -> Option<GuideExon> {
        let tree = self.tree.as_ref()?;
        if qstart == 0 && qend == 0 {
            return None;
        }

        let q_last = qend.saturating_sub(1) as i32;
        let mut found: Option<GuideExon> = None;

        tree.query(qstart as i32, q_last, |node| {
            if found.is_some() {
                return;
            }
            if node.metadata.tid == tid {
                let s = node.first.max(0) as u32;
                let e_excl = (node.last + 1).max(0) as u32;
                let data = &node.metadata;
                found = Some(GuideExon {
                    start: s,
                    end: e_excl,
                    pos: 0,
                    pos_start: data.pos_start,
                    exon_id: data.exon_id,
                    seq: self
                        .seqs
                        .get(&(data.tid, data.exon_id))
                        .map(Arc::clone)
                        .unwrap_or_else(|| Arc::from([])),
                    left_ins: 0,
                    right_ins: 0,
                    left_gap: 0,
                    right_gap: 0,
                    has_prev: data.has_prev,
                    has_next: data.has_next,
                    prev_start: data.prev_start,
                    prev_end: data.prev_end,
                    next_start: data.next_start,
                    next_end: data.next_end,
                    transcript_len: data.transcript_len,
                });
            }
        });

        found
    }

    pub fn find_overlapping(
        &self,
        qstart: u32,
        qend: u32,
        strand: char,
        config: &ReadEvaluationConfig,
        status: ExonStatus,
        out: &mut HashMap<Tid, GuideExon>,
    ) {
        out.clear();
        let tree = match self.tree.as_ref() {
            Some(t) => t,
            None => return,
        };

        let q_last = qend.saturating_sub(1) as i32;

        tree.query(qstart as i32, q_last, |node| {
            let s = node.first.max(0) as u32;
            let e_excl = (node.last + 1).max(0) as u32;

            if (!config.soft_clips || config.strict) && (qstart < s || e_excl < qend) {
                return;
            }

            let data = &node.metadata;
            let mut exon = GuideExon {
                start: s,
                end: e_excl,
                pos: 0,
                pos_start: data.pos_start,
                exon_id: data.exon_id,
                seq: self
                    .seqs
                    .get(&(data.tid, data.exon_id))
                    .map(Arc::clone)
                    .unwrap_or_else(|| Arc::from([])),
                left_ins: 0,
                right_ins: 0,
                left_gap: 0,
                right_gap: 0,
                has_prev: data.has_prev,
                has_next: data.has_next,
                prev_start: data.prev_start,
                prev_end: data.prev_end,
                next_start: data.next_start,
                next_end: data.next_end,
                transcript_len: data.transcript_len,
            };

            let mut rejected = false;

            if strand == '+' {
                if s <= qstart {
                    exon.pos = (qstart - s) + data.pos_start;
                    exon.left_gap = (qstart - s) as i32;
                    if matches!(status, ExonStatus::MiddleExon | ExonStatus::LastExon)
                        && (exon.left_gap as u32) > config.max_junc_gap
                    {
                        rejected = true;
                    }
                } else {
                    exon.pos = data.pos_start;
                    exon.left_ins = (s - qstart) as i32;
                    if matches!(status, ExonStatus::MiddleExon | ExonStatus::LastExon) {
                        if (exon.left_ins as u32) > config.max_junc_ins {
                            rejected = true;
                        }
                    } else if (exon.left_ins as u32) > config.max_clip {
                        rejected = true;
                    }
                }

                if !rejected {
                    if e_excl < qend {
                        exon.right_ins = (qend - e_excl) as i32;
                        if matches!(status, ExonStatus::FirstExon | ExonStatus::MiddleExon) {
                            if (exon.right_ins as u32) > config.max_junc_ins {
                                rejected = true;
                            }
                        } else if (exon.right_ins as u32) > config.max_clip {
                            rejected = true;
                        }
                    } else if qend < e_excl {
                        exon.right_gap = (e_excl - qend) as i32;
                        if matches!(status, ExonStatus::FirstExon | ExonStatus::MiddleExon)
                            && (exon.right_gap as u32) > config.max_junc_gap
                        {
                            rejected = true;
                        }
                    }
                }
            } else {
                if qend <= e_excl {
                    exon.pos = (e_excl - qend) + data.pos_start;
                    exon.right_gap = (e_excl - qend) as i32;
                    if matches!(status, ExonStatus::FirstExon | ExonStatus::MiddleExon)
                        && (exon.right_gap as u32) > config.max_junc_gap
                    {
                        rejected = true;
                    }
                } else {
                    exon.pos = data.pos_start;
                    exon.right_ins = (qend - e_excl) as i32;
                    // NOTE: Match C++ behavior (g2t.cpp) which effectively
                    // treats this condition as always true due to `|| MIDDLE_EXON`.
                    // That means `max_junc_ins` is enforced for all statuses here.
                    if (exon.right_ins as u32) > config.max_junc_ins {
                        rejected = true;
                    }
                }

                if !rejected {
                    if qstart < s {
                        exon.left_ins = (s - qstart) as i32;
                        if matches!(status, ExonStatus::MiddleExon | ExonStatus::LastExon) {
                            if (exon.left_ins as u32) > config.max_junc_ins {
                                rejected = true;
                            }
                        } else if (exon.left_ins as u32) > config.max_clip {
                            rejected = true;
                        }
                    } else if s < qstart {
                        exon.left_gap = (qstart - s) as i32;
                        if matches!(status, ExonStatus::MiddleExon | ExonStatus::LastExon)
                            && (exon.left_gap as u32) > config.max_junc_gap
                        {
                            rejected = true;
                        }
                    }
                }
            }

            if !rejected {
                out.insert(data.tid, exon);
            }
        });
    }
}

impl Default for IntervalTree {
    fn default() -> Self {
        Self::new()
    }
}

pub struct G2TTree {
    // Per refid: (forward, reverse)
    trees: HashMap<RefId, (IntervalTree, IntervalTree)>,
    name_id_map: HashMap<Tid, String>,
    tid_names: Vec<String>,
    // Transcript length (sum of exon lengths) indexed by `Tid`.
    tid_lengths: Vec<u32>,
}

impl G2TTree {
    pub fn new() -> Self {
        Self {
            trees: HashMap::new(),
            name_id_map: HashMap::new(),
            tid_names: Vec::new(),
            tid_lengths: Vec::new(),
        }
    }

    /// Number of transcripts indexed in this tree.
    pub fn num_transcripts(&self) -> usize {
        self.tid_names.len()
    }

    /// The name of the transcript with the given `tid`, if present.
    pub fn transcript_name(&self, tid: Tid) -> Option<&str> {
        self.tid_names.get(tid as usize).map(|s| s.as_str())
    }

    /// All transcript names, indexed by `tid` (dense, build order).
    pub fn transcript_names(&self) -> &[String] {
        &self.tid_names
    }

    /// The length (sum of exon lengths) of the transcript with the given `tid`.
    pub fn transcript_len(&self, tid: Tid) -> Option<u32> {
        self.tid_lengths.get(tid as usize).copied()
    }

    /// All transcript lengths, indexed by `tid` (dense, build order).
    pub fn transcript_lengths(&self) -> &[u32] {
        &self.tid_lengths
    }

    /// Record the length (sum of exon lengths) for the transcript with the
    /// given `tid`. Used by [`build_g2t`] once the full exon chain is known.
    pub fn set_transcript_len(&mut self, tid: Tid, len: u32) {
        let idx = tid as usize;
        if idx >= self.tid_lengths.len() {
            self.tid_lengths.resize(idx + 1, 0);
        }
        self.tid_lengths[idx] = len;
    }

    fn get_trees_mut(&mut self, refid: RefId) -> &mut (IntervalTree, IntervalTree) {
        self.trees.entry(refid).or_insert_with(|| {
            (IntervalTree::new(), IntervalTree::new())
        })
    }

    fn get_tree_mut(&mut self, refid: RefId, strand: char) -> &mut IntervalTree {
        let (fw, rc) = self.get_trees_mut(refid);
        if strand == '-' {
            rc
        } else {
            fw
        }
    }

    pub fn insert_tid_name(&mut self, tid: Tid, tid_name: &str) {
        if let std::collections::hash_map::Entry::Vacant(entry) = self.name_id_map.entry(tid) {
            entry.insert(tid_name.to_string());
            self.tid_names.push(tid_name.to_string());
        }
    }

    pub fn add_interval(
        &mut self,
        refid: RefId,
        tid: Tid,
        interval: &IntervalData,
        strand: char,
    ) {
        let tree = self.get_tree_mut(refid, strand);
        tree.add_interval(tid, interval);
    }

    pub fn index_ref(&mut self, refid: RefId) {
        let (fw, rc) = self.get_trees_mut(refid);
        fw.index();
        rc.index();
    }

    pub fn get_guide_exon_for_tid(
        &self,
        refid: RefId,
        strand: char,
        tid: Tid,
        start: u32,
        end: u32,
    ) -> Option<GuideExon> {
        let tree = self.get_tree(refid, strand)?;
        tree.find_overlapping_for_tid(start, end, tid)
    }

    #[allow(clippy::too_many_arguments)]
    pub fn get_guide_exons(
        &self,
        refid: RefId,
        strand: char,
        start: u32,
        end: u32,
        config: &ReadEvaluationConfig,
        status: ExonStatus,
        out: &mut HashMap<Tid, GuideExon>,
    ) {
        if let Some(tree) = self.get_tree(refid, strand) {
            tree.find_overlapping(start, end, strand, config, status, out);
        } else {
            out.clear();
        }
    }

    fn get_tree(&self, refid: RefId, strand: char) -> Option<&IntervalTree> {
        let pair = self.trees.get(&refid)?;
        if strand == '-' {
            Some(&pair.1)
        } else {
            Some(&pair.0)
        }
    }
}

impl Default for G2TTree {
    fn default() -> Self {
        Self::new()
    }
}

pub fn build_g2t(
    transcripts: &[Transcript],
    refname_to_id: &HashMap<String, RefId>,
    fasta: Option<&FastaDb>,
) -> Result<G2TTree> {
    let mut g2t = G2TTree::new();
    let mut touched_refids: HashSet<RefId> = HashSet::new();

    // Assign tid incrementally in stable order of transcripts slice
    for (tid_idx, tx) in transcripts.iter().enumerate() {
        let tid = tid_idx as Tid;
        g2t.insert_tid_name(tid, &tx.id);

        let refid = *refname_to_id
            .get(&tx.seqname)
            .ok_or_else(|| anyhow!("reference '{}' not found in BAM header", tx.seqname))?;
        touched_refids.insert(refid);

        let mut exons: Vec<Exon> = tx.exons.clone();
        exons.sort_by_key(|e| e.start);

        let mut pos_start: u32 = 0;
        let mut intervals: Vec<IntervalData> = Vec::with_capacity(exons.len());
        let exon_iter: Box<dyn Iterator<Item = (usize, &Exon)>> = if tx.strand == '-' {
            Box::new(exons.iter().enumerate().rev())
        } else {
            Box::new(exons.iter().enumerate())
        };

        for (idx, exon) in exon_iter {
            let len = exon.end.saturating_sub(exon.start);
            let seq: Arc<[u8]> = if let Some(fa) = fasta {
                fa.get_slice(&tx.seqname, exon.start, exon.end)
                    .map(Arc::from)
                    .unwrap_or_else(|| Arc::from([]))
            } else {
                Arc::from([])
            };
            intervals.push(IntervalData {
                start: exon.start,
                end: exon.end,
                idx: idx as u8,
                pos_start,
                seq,
                has_prev: false,
                has_next: false,
                prev_start: 0,
                prev_end: 0,
                next_start: 0,
                next_end: 0,
                transcript_len: 0,
            });
            pos_start = pos_start.saturating_add(len);
        }

        let transcript_len = pos_start;
        g2t.set_transcript_len(tid, transcript_len);

        for k in 0..intervals.len() {
            let mut interval = intervals[k].clone();
            if k > 0 {
                interval.has_prev = true;
                interval.prev_start = intervals[k - 1].start;
                interval.prev_end = intervals[k - 1].end;
            }
            if k + 1 < intervals.len() {
                interval.has_next = true;
                interval.next_start = intervals[k + 1].start;
                interval.next_end = intervals[k + 1].end;
            }
            interval.transcript_len = transcript_len;

            g2t.add_interval(refid, tid, &interval, tx.strand);
        }

    }

    // Index all trees only after all intervals are inserted.
    for refid in touched_refids {
        g2t.index_ref(refid);
    }

    Ok(g2t)
}

/// Convenience wrapper around [`build_g2t`] that takes the reference (chromosome)
/// names as a slice, assigning each its position in the slice as its `RefId`.
///
/// This lets callers build the index from, e.g., a BAM header's reference list
/// or a minimap2 target list without constructing the internal
/// `HashMap<String, RefId>` themselves. The resulting `RefId` for a chromosome
/// is its 0-based index in `refnames`, which must match the `ref_id` set on the
/// [`GenomicAlignment`](crate::GenomicAlignment)s passed to
/// [`project_group`](crate::project_group).
pub fn build_g2t_from_refnames(
    transcripts: &[Transcript],
    refnames: &[String],
    fasta: Option<&FastaDb>,
) -> Result<G2TTree> {
    let mut refname_to_id: HashMap<String, RefId> = HashMap::with_capacity(refnames.len());
    for (idx, name) in refnames.iter().enumerate() {
        refname_to_id.insert(name.clone(), idx as RefId);
    }
    build_g2t(transcripts, &refname_to_id, fasta)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::annotation::{Exon, Transcript};

    /// `build_g2t` must expose dense transcript ids, names, and lengths (sum of
    /// exon lengths) so library consumers (e.g. oarfish) can build a
    /// transcriptome header without re-reading the annotation.
    #[test]
    fn transcript_names_and_lengths_are_exposed() {
        // tx1: chr1 '+', two exons [100,150) and [200,250) -> length 100.
        // tx2: chr1 '-', one exon [300,330) -> length 30.
        let txs = vec![
            Transcript {
                id: "tx1".to_string(),
                seqname: "chr1".to_string(),
                strand: '+',
                exons: vec![Exon { start: 100, end: 150 }, Exon { start: 200, end: 250 }],
            },
            Transcript {
                id: "tx2".to_string(),
                seqname: "chr1".to_string(),
                strand: '-',
                exons: vec![Exon { start: 300, end: 330 }],
            },
        ];
        let refnames = vec!["chr1".to_string()];
        let g2t = build_g2t_from_refnames(&txs, &refnames, None).expect("build_g2t");

        assert_eq!(g2t.num_transcripts(), 2);
        assert_eq!(g2t.transcript_name(0), Some("tx1"));
        assert_eq!(g2t.transcript_name(1), Some("tx2"));
        assert_eq!(g2t.transcript_name(2), None);
        assert_eq!(g2t.transcript_len(0), Some(100));
        assert_eq!(g2t.transcript_len(1), Some(30));
        assert_eq!(g2t.transcript_lengths(), &[100, 30]);
        assert_eq!(g2t.transcript_names(), &["tx1".to_string(), "tx2".to_string()]);
    }
}
